[torch][segment_reduce] Add cuda support for mean reduction

serhaty opened the pull request:
Summary:
Building on top of previous PR: https://github.com/pytorch/pytorch/pull/59521

This diff is adding support for mean reduction for Cuda (fwd only currently).
Will add cuda backward implementation in subsequent PR.
Next Steps:
cuda backward support for mean
2d data input support
more testing
benchmarking

Test Plan: update unit test to cover this part as well.

Differential Revision: D28922838

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -10034,7 +10034,7 @@
     CPU, CUDA: segment_reduce_kernel
 
-- func: segment_reduce_backward(Tensor grad, Tensor output, Tensor data, *, Tensor? lengths=None) -> Tensor
+- func: segment_reduce_backward(Tensor grad, Tensor output, Tensor data, str reduce, *, Tensor? lengths=None) -> Tensor
```
also bc breaking

ngimel commented on aten/src/ATen/native/SegmentReduce.cpp:
```diff
@@ -29,20 +40,41 @@ Tensor _segment_reduce_cpu_kernel(
+
+          // ===== step3: finalize reduction
+          TORCH_CHECK(lengths_data[i] >= 0);
+          if (lengths_data[i] == 0 && !initial.has_value()) {
```
you don't check for SegmentReductionType here? Would make sense to add a test to catch that

serhaty replied:
if it is 0 length segment and there is no initial provided, we should always set to NAN, regardless of aggregation type, right?

ngimel replied:
I don't know, numpy errors out on max reduction over 0 elements, I guess it's reasonable to either error out or return -infinity for max reduction

serhaty replied:
Discussed offline, will change defaults at the top of the stack.

ngimel commented on aten/src/ATen/native/cuda/SegmentReduce.cu:
```diff
@@ -61,20 +93,36 @@ Tensor _segment_reduce_cuda_kernel(
+              initial_value,
+              at::cuda::getCurrentCUDAStream());
+        } else if (reduction == SegmentReductionType::MEAN) {
+          scalar_t initial_value =
```
It would be better to use cub here also, and divide result by lengths. You can put special initial value handling in the kernel dividing by lengths if you want.

serhaty replied:
sounds good!

ngimel commented on aten/src/ATen/native/cuda/SegmentReduce.cu:
```diff
@@ -52,6 +79,11 @@ Tensor _segment_reduce_cuda_kernel(
   auto offsets = _get_complete_sum(lengths);
```
I forgot, is it checked somewhere that lengths are contiguous? Otherwise this won't work.

serhaty replied:
yes, caller on .cpp file is doing both for cuda and cpu:

https://github.com/pytorch/pytorch/blob/0054702491e7ba8fcdc0ff7b87760782497841cc/aten/src/ATen/native/SegmentReduce.cpp#L180

ngimel commented on aten/src/ATen/native/cuda/SegmentReduce.cu:
```diff
@@ -23,6 +23,19 @@ struct CustomMax {
   }
 };
 
+struct CustomSum {
```
do you need special nan handling here? `a+b` should propagate nans.

serhaty replied:
Given the last time issues (this part and also making it work for rocm) with this for max, just followed safe path that I know works.

ngimel replied:
`max` is a known bug on rocm, sum is a standard operation

ngimel commented on aten/src/ATen/native/cuda/SegmentReduce.cu:
```diff
@@ -53,6 +87,11 @@ Tensor _segment_reduce_cuda_kernel(
   auto* offsets_data_ptr = offsets.data_ptr<int64_t>();
 
+  constexpr int threads_per_block = 256;
+  int num_blocks = segment_count / threads_per_block;
```
this will not correctly compute num_blocks if `segment_count` is not evenly divisible by `threads_per_block`, so you'd be needlessly doing grid-stride loop. It's better to do `int64_t num_blocks = (segment_count + threads_per_block - 1)/threads_per_block`, and then you can remove grid-stride loop. You can have an assert that segment_count is less than max_blocks * 256, but in reality it'll never fire, max_blocks is 2**31.

ngimel reviewed (commented):
The general structure is good, minor comments.

Commit:
[torch][segment_reduce] Add cuda support for mean reduction (#59543)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/59543

Building on top of previous PR: https://github.com/pytorch/pytorch/pull/59521

This diff is adding support for mean reduction for Cuda (fwd only currently).
Will add cuda backward implementation in subsequent PR.
Next Steps:
cuda backward support for mean
2d data input support
more testing
benchmarking

Test Plan: update unit test to cover this part as well.

Differential Revision: D28922838

fbshipit-source-id: 14df541e341e73eb32fe08730315d8131d8592c1

ngimel commented on aten/src/ATen/native/cuda/SegmentReduce.cu:
```diff
@@ -23,6 +23,14 @@ struct CustomMax {
   }
 };
 
+struct CustomSum {
```
nit: it does not need to be called customSum, just Sum?
